src/groupeddataframe/splitapplycombine.jl

```diff
@@ -523,58 +536,62 @@ function _combine(gd::GroupedDataFrame,
     # and if a given column can be replaced in the future
     seen_cols = Dict{Symbol, Tuple{Bool, Int}}()
 
+    tasks = similar(cs_norm, Task)
+
     parentdf = parent(gd)
-    for i in eachindex(cs_norm, optional_transform)
+    for i in eachindex(cs_norm, optional_transform, tasks)
         cs_i = cs_norm[i]
         optional_i = optional_transform[i]
 
-        if length(gd) > 0 && isagg(cs_i, gd)
-            _combine_process_agg(cs_i, optional_i, parentdf, gd, seen_cols, trans_res, idx_agg)
+        tasks[i] = @spawn if length(gd) > 0 && isagg(cs_i, gd)
+            _combine_process_agg(cs_i, optional_i, parentdf, gd, seen_cols, trans_res, idx_agg[])
         elseif keeprows && cs_i isa Pair && first(last(cs_i)) === identity &&
                !(first(cs_i) isa AsTable) && (last(last(cs_i)) isa Symbol)
             # this is a fast path used when we pass a column or rename a column in select or transform
             _combine_process_noop(cs_i, optional_i, parentdf, seen_cols, trans_res, idx_keeprows, copycols)
         elseif cs_i isa Base.Callable
-            idx_callable = _combine_process_callable(cs_i, optional_i, parentdf, gd, seen_cols, trans_res, idx_agg)
-            if idx_callable !== nothing
-                if idx_agg === nothing
-                    idx_agg = idx_callable
-                else
-                    @assert idx_agg === idx_callable
-                end
-            end
+            _combine_process_callable(cs_i, optional_i, parentdf, gd, seen_cols, trans_res, idx_agg)
         else
             @assert cs_i isa Pair
-            idx_pair = _combine_process_pair(cs_i, optional_i, parentdf, gd, seen_cols, trans_res, idx_agg)
-            if idx_pair !== nothing
-                if idx_agg === nothing
-                    idx_agg = idx_pair
-                else
-                    @assert idx_agg === idx_pair
-                end
-            end
+            _combine_process_pair(cs_i, optional_i, parentdf, gd, seen_cols, trans_res, idx_agg)
         end
     end
+    # Workaround JuliaLang/julia#38931:
+    # we want to preserve the exception type thrown in user code,
+    # and print the backtrace corresponding to it
+    for t in tasks
+        try
+            wait(t)
+        catch e
+            throw(t.exception)
```

Comment:
maybe add `@assert e isa TaskFailedException`. This is consistent with `wait` docstring.

Actually I would say that it is needed to do so, as otherwise we have:
```
julia> Threads.@spawn sqrt(-1)
Task (failed) @0x00007f5ceb93b0d0
DomainError with -1.0:
sqrt will only return a complex result if called with a complex argument. Try sqrt(Complex(x)).
throw_complex_domainerror(::Symbol, ::Float64) at ./math.jl:33
sqrt at ./math.jl:573 [inlined]
sqrt(::Int64) at ./math.jl:599
(::var"#11#12")() at ./threadingconstructs.jl:169

julia> t = Threads.@spawn begin
       sleep(5)
       sqrt(-1)
       end
Task (runnable) @0x00007f5cebce24a0

julia> wait(t)
ERROR: TaskFailedException:
DomainError with -1.0:
sqrt will only return a complex result if called with a complex argument. Try sqrt(Complex(x)).
Stacktrace:
 [1] throw_complex_domainerror(::Symbol, ::Float64) at ./math.jl:33
 [2] sqrt at ./math.jl:573 [inlined]
 [3] sqrt(::Int64) at ./math.jl:599
 [4] macro expansion at ./REPL[10]:3 [inlined]
 [5] (::var"#13#14")() at ./threadingconstructs.jl:169
Stacktrace:
 [1] wait(::Task) at ./task.jl:267
 [2] top-level scope at REPL[11]:1
```

Reply:
Good idea. Actually we can even be more general and call `rethrow(e)` if we get an unexpected type, that can't hurt.

What do you mean with this example? Just that the exception type isn't `DomainError` as one would expect?

Reply:
By this example I mean that when you do `@spawn` you might get an exception before `wait` is called.

So in general - you could have two different exception types thrown from some code (as what I passed in `begin`-`end` block could be anything - either something that throws immediately or something that will throw only when `wait` is called)